include/Target/Btor/BtorToBtorIRTranslation.h

```diff
@@ -0,0 +1,190 @@
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+  
+  std::map<int64_t, Value> cache;
+
+  OwningOpRef<FuncOp> buildInitFunction();
+  OwningOpRef<FuncOp> buildNextFunction();
+  
+ private: 
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  Btor2Parser *model = nullptr;
+  FILE *modelFile = nullptr;
+
+  void parseModelLine(Btor2Line *l);
+
+  void filterInits(){
+    size_t i = 0;
+    for (size_t j = 0, sz = inits.size(); j < sz; ++j) {
+        if (inits.at(j)) {
+            inits[i++] = inits.at(j);
+        }
+    }
+    inits.resize(i);
+  }
+
+  void filterNexts() {
+    size_t i = 0;
+    for (size_t j = 0, sz = nexts.size(); j < sz; ++j) {
+      if (nexts.at(j)) {
+        nexts[i++] = nexts.at(j);
+      }
+    }
+    nexts.resize(i);
+  }
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+
+  MLIRContext *context;
+  OpBuilder builder;
+  Location unknownLoc;
+  
+  void toOp(Btor2Line *line);
+  bool isValidChild(Btor2Line * line);
+  void createNegateLine(int64_t curAt, Value child);
+  Operation * createMLIR(const Btor2Line *line, const int64_t *kids);
+
+  template <typename btorOp>
+  Operation * buildBinaryOp(Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, lhs, rhs);
+    return res;
+  }
+
+  template <typename btorOp>
+  Operation * buildComparisonOp(btor::BtorPredicate pred,
+                                Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, pred, lhs, rhs);
+    return res;
+  }
+
+  template <typename btorOp>
+  Operation * buildOverflowOp(Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, 
+                                    builder.getIntegerType(1), 
+                                    lhs, rhs);
+    return res;
+  }
+
+  Operation * buildConstantOp(uint32_t width, std::string str, uint32_t radix) {
+    Type type = builder.getIntegerType(width);
+    mlir::APInt value(width, 0, radix);
+    if (str.compare("ones") == 0) {
+      value.setAllBits();
+    } else if (str.compare("one") == 0) {
+      value = mlir::APInt(width, 1, radix);
+    } else if (str.compare("zero") != 0) {
+      value = mlir::APInt(width, str, radix);
+    }
+    auto res = builder.create<btor::ConstantOp>(unknownLoc, type,
+                        builder.getIntegerAttr(type, value.getLimitedValue()));
+    return res;
+  }
+
+  Operation * buildConcatOp(Value lhs, Value rhs) {
+    auto newWidth = lhs.getType().getIntOrFloatBitWidth() +
+               rhs.getType().getIntOrFloatBitWidth();
+    Type resType = builder.getIntegerType(newWidth);
```

Comment:
there are quite a few things that are being passed by value, here it is `Type`. Perhaps that is the convention that MLIR is using. I will stop commenting about this for now.